File: backend/src/services/activityLogger.js
const pool = require('../config/db');

/**
 * Log an activity to the activity_logs table
 * @param {object} params
 * @param {string} params.entity_type - e.g. 'stock_transfer', 'procurement', 'dispatch', 'receive', 'request', 'consumable'
 * @param {number} params.entity_id - The ID of the entity being acted upon
 * @param {string} params.action - e.g. 'created', 'updated', 'deleted', 'approved', 'rejected', 'transferred', 'dispatched', 'received', 'status_changed'
 * @param {string} [params.details] - Human-readable summary
 * @param {object} [params.changes] - JSON object describing before/after changes
 * @param {string} params.performed_by - Name of the user who performed the action
 */
const logActivity = async ({ entity_type, entity_id, action, details, changes, performed_by }) => {
  try {
    await pool.query(
      `INSERT INTO activity_logs (entity_type, entity_id, action, details, changes, performed_by)
       VALUES ($1, $2, $3, $4, $5, $6)`,
      [entity_type, entity_id, action, details || '', JSON.stringify(changes || {}), performed_by]
    );
  } catch (err) {
    console.error('[ActivityLogger] Failed to log activity:', err.message);
  }
};

module.exports = { logActivity };

File: backend/app/services/activity_logger.py
import json

from app.db import query


def log_activity(entity_type, entity_id, action, performed_by, details="", changes=None):
    """Log an activity to the activity_logs table. Never raises — mirrors the Node fire-and-forget logger."""
    try:
        query(
            """INSERT INTO activity_logs (entity_type, entity_id, action, details, changes, performed_by)
               VALUES (%s, %s, %s, %s, %s, %s)""",
            [entity_type, entity_id, action, details or "", json.dumps(changes or {}), performed_by],
        )
    except Exception as err:
        print(f"[ActivityLogger] Failed to log activity: {err}")
